kernel/sleep/src/lib.rs

```diff
@@ -9,128 +9,219 @@
 //! TODO: use regular time-keeping abstractions like Duration and Instant.
 
 #![no_std]
+#![feature(let_else)]
+
 extern crate task;
 extern crate irq_safety;
 extern crate alloc;
 #[macro_use] extern crate lazy_static;
 extern crate scheduler;
 
-use core::sync::atomic::{Ordering, AtomicUsize};
 use alloc::collections::binary_heap::BinaryHeap;
+use alloc::sync::Arc;
+use core::cmp;
+use core::future::Future;
+use core::pin::Pin;
+use core::sync::atomic::{Ordering, AtomicUsize};
+use core::task::{Context, Poll, Waker};
+
 use irq_safety::MutexIrqSafe;
-use task::{current_task, TaskRef};
 
-/// Contains the `TaskRef` and the associated wakeup time for an entry in DELAYED_TASKLIST.
-#[derive(Clone, Eq, PartialEq)]
-struct SleepingTaskNode {
-    resume_time: usize,
-    taskref: TaskRef,
+/// A duration given in ticks of the [APIC timer].
+/// 
+/// [APIC timer]: interrupts::lapic_timer_handler
+pub type Ticks = usize;
+
+pub type AtomicTicks = AtomicUsize;
```

Comment:
```suggestion
pub type Ticks = u64;

pub type AtomicTicks = AtomicU64
```
I know ticks were stored as a `usize` before this pull request, but might as well comment on it now. `usize`s are generally used for computation related to memory, so ticks should be a `u64` (or `u32`?). It doesn't make sense for the type of `TICK_COUNT` to change based on the target architecture.

Also, more specific to this PR, is it really necessary to use type aliases here?

Reply:
Agree, it deserves a comment. I had actually made it `u64` but then reverted when I looked into atomics more and saw that `AtomicU64` isn't supported on all platforms. `AtomicUsize` is the most portable atomic int. I presume that's why the original code used it.

Reply:
Do you not like the type aliases? I was actually considering going one step further and making them newtypes, and/or using `Duration`s from a time crate. I don't like how the values lack units.

Reply:
I guess it's a matter of style. The issue with using concrete units (i.e. `Duration) is that the LAPIC timer's frequency is dependent on the CPU's frequency, so the tick interval varies from machine to machine.